daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Anf.scala

```diff
@@ -284,156 +277,128 @@ private[lf] object Anf {
     *  Note: this wrapping is the reason why we need a "second" CPS transform to
     *  achieve constant stack through trampoline.
     */
-  private[this] def transformExp[A](
-      depth: DepthA,
-      env: Env,
-      exp: source.SExpr,
-      k: K[target.SExpr, A],
-  )(
-      transform: Tx[target.SExpr, A]
-  ): Trampoline[A] =
+  private[this] def transformExp(depth: DepthA, env: Env, exp: source.SExpr)(
+      k: K[target.SExpr]
+  )(transform: Tx[target.SExpr]): Res = Bounce { () =>
     exp match {
       case atom0: source.SExprAtomic =>
         val atom = makeRelativeA(depth)(makeAbsoluteA(env, atom0))
-        Bounce(() => transform(depth, atom, k))
+        transform(depth, atom)(k)
 
-      case source.SEVal(x) => Bounce(() => transform(depth, target.SEVal(x), k))
+      case source.SEVal(x) => transform(depth, target.SEVal(x))(k)
 
       case source.SEApp(func, args) =>
         // It's safe to perform ANF if the func-expression has no effects when evaluated.
         val safeFunc =
           func match {
             // we know that trivially in these two cases
-            case source.SEBuiltin(b) => (args.size <= b.arity)
+            case source.SEBuiltin(b) =>
+              val overApp = args.lengthCompare(b.arity) > 0
+              !overApp
             case _ => false
           }
         // It's also safe to perform ANF for applications of a single argument.
-        if (safeFunc || args.size == 1) {
-          transformMultiApp[A](depth, env, func, args.toArray, k)(transform)
+        val singleArg = args.lengthCompare(1) == 0
```

Comment:
If you want to use a additional `val`, you may use simple pattern matching.
More efficient and maybe easier to understand for non scala experts. 
```suggestion
        val singleArg = args match {
          case _ :: Nil => true 
          case _ => false 
        }
```
Just a suggestion, I let you choose what you think is the best.